components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java

```diff
@@ -184,13 +185,28 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
         // Update Service Provider
         ServiceProvider sp = getServiceProvider(appDTO.getApplicationName(), tenantDomain);
         if (StringUtils.isNotEmpty(clientName)) {
+            // Check whether a service provider already exists for the name we are trying
+            // to register the OAuth app with.
+            if (!appDTO.getApplicationName().equals(clientName) && isServiceProviderExist(clientName, tenantDomain)) {
+                throw DCRMUtils.generateClientException(DCRMConstants.ErrorMessages.CONFLICT_EXISTING_APPLICATION,
+                        clientName);
+            }
+
             // Regex validation of the application name.
             if (!DCRMUtils.isRegexValidated(clientName)) {
                 throw DCRMUtils.generateClientException(DCRMConstants.ErrorMessages.BAD_REQUEST_INVALID_SP_NAME,
                         DCRMUtils.getSPValidatorRegex(), null);
             }
-            sp.setApplicationName(clientName);
-            updateServiceProvider(sp, tenantDomain, applicationOwner);
+            if (sp == null) {
+                throw DCRMUtils.generateClientException(DCRMConstants.ErrorMessages.FAILED_TO_GET_SP,
+                        appDTO.getApplicationName(), null);
+            }
+            // Need to create a deep clone, since modifying the fields of the original object,
+            // will modify the cached SP object.
+            Gson gson = new Gson();
```

Comment:
Better to make the Gson clone logic to a private method

private ServiceProvider  cloneServiceProvider  (ServiceProvider original) {...}

Reply:
Fixed https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1527/commits/d2d458049c32fdfb7b4a360826a60165c79f04c4